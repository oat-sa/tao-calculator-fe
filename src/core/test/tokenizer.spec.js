/**
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; under version 2
 * of the License (non-upgradable).
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *
 * Copyright (c) 2018-2023 (original work) Open Assessment Technologies SA ;
 */

import tokenizerFactory from '../tokenizer.js';
import terms from '../terms.js';

describe('tokenizer', () => {
    it('is a factory', () => {
        expect(tokenizerFactory).toEqual(expect.any(Function));
        expect(tokenizerFactory()).toEqual(expect.any(Object));
        expect(tokenizerFactory()).not.toStrictEqual(tokenizerFactory());
    });

    describe('tokenize', () => {
        it('is a function', () => {
            expect(tokenizerFactory().tokenize).toEqual(expect.any(Function));
        });

        it.each([
            ['(.1 + .2) * 10^8 + 4 @nthrt 8e5', void 0],
            [' 3+4 *$foo + sinh 1', void 0],
            [
                ' 3+4 *$foo + sinh PI',
                {
                    symbols: {
                        DOLLAR: '$'
                    },
                    keywords: {
                        FOO: 'foo'
                    }
                }
            ]
        ])('extracts tokens from the expression "%s"', (expression, config) => {
            const tokenizer = tokenizerFactory(config);
            const tokens = tokenizer.tokenize(expression);

            expect(tokens).toMatchSnapshot();
        });

        it('recognize all the registered terms', () => {
            const expression = Object.keys(terms).reduce((expr, id) => `${expr} ${terms[id].value}`, '');

            const tokenizer = tokenizerFactory();
            const tokens = tokenizer.tokenize(expression);

            expect(tokens).toMatchSnapshot();
        });
    });

    describe('iterator', () => {
        it('is a function', () => {
            expect(tokenizerFactory().iterator).toEqual(expect.any(Function));
        });

        it.each([
            ['(.1 + .2) * 10^8 + 4 @nthrt 8e5', void 0],
            [' 3+4 *$foo + sinh 1', void 0],
            [
                ' 3+4 *$foo + sinh PI',
                {
                    symbols: {
                        DOLLAR: '$'
                    },
                    keywords: {
                        FOO: 'foo'
                    }
                }
            ]
        ])('iterates over the terms from the expression "%s"', (expression, config) => {
            const tokenizer = tokenizerFactory(config);
            const next = tokenizer.iterator(expression);

            expect(next).toEqual(expect.any(Function));

            const tokens = [];
            for (let token = next(); token; token = next()) {
                tokens.push(token);
            }

            expect(tokens).toMatchSnapshot();
        });
    });
});
