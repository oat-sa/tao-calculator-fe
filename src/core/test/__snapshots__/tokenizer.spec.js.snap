// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tokenizer iterator iterates over the terms from the expression " 3+4 *$foo + sinh 1" 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "3",
    "toString": [Function],
    "type": "NUM3",
    "value": "3",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "+",
    "toString": [Function],
    "type": "POS",
    "value": "+",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "4",
    "toString": [Function],
    "type": "NUM4",
    "value": "4",
  },
  {
    "col": 6,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": "*",
    "toString": [Function],
    "type": "MUL",
    "value": "*",
  },
  {
    "col": 7,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": "$foo + sinh 1",
    "toString": [Function],
    "type": "syntaxError",
    "value": "$foo + sinh 1",
  },
]
`;

exports[`tokenizer iterator iterates over the terms from the expression " 3+4 *$foo + sinh PI" 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "3",
    "toString": [Function],
    "type": "NUM3",
    "value": "3",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "+",
    "toString": [Function],
    "type": "POS",
    "value": "+",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "4",
    "toString": [Function],
    "type": "NUM4",
    "value": "4",
  },
  {
    "col": 6,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": "*",
    "toString": [Function],
    "type": "MUL",
    "value": "*",
  },
  {
    "col": 7,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": "$",
    "toString": [Function],
    "type": "DOLLAR",
    "value": "$",
  },
  {
    "col": 8,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": "foo",
    "toString": [Function],
    "type": "FOO",
    "value": "foo",
  },
  {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 14,
    "line": 1,
    "lineBreaks": 0,
    "offset": 13,
    "text": "sinh",
    "toString": [Function],
    "type": "SINH",
    "value": "sinh",
  },
  {
    "col": 19,
    "line": 1,
    "lineBreaks": 0,
    "offset": 18,
    "text": "PI",
    "toString": [Function],
    "type": "PI",
    "value": "PI",
  },
]
`;

exports[`tokenizer iterator iterates over the terms from the expression "(.1 + .2) * 10^8 + 4 @nthrt 8e5" 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "(",
    "toString": [Function],
    "type": "LPAR",
    "value": "(",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": ".",
    "toString": [Function],
    "type": "DOT",
    "value": ".",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "1",
    "toString": [Function],
    "type": "NUM1",
    "value": "1",
  },
  {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": ".",
    "toString": [Function],
    "type": "DOT",
    "value": ".",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": "2",
    "toString": [Function],
    "type": "NUM2",
    "value": "2",
  },
  {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": ")",
    "toString": [Function],
    "type": "RPAR",
    "value": ")",
  },
  {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": "*",
    "toString": [Function],
    "type": "MUL",
    "value": "*",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 12,
    "text": "1",
    "toString": [Function],
    "type": "NUM1",
    "value": "1",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 13,
    "text": "0",
    "toString": [Function],
    "type": "NUM0",
    "value": "0",
  },
  {
    "col": 15,
    "line": 1,
    "lineBreaks": 0,
    "offset": 14,
    "text": "^",
    "toString": [Function],
    "type": "POW",
    "value": "^",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 15,
    "text": "8",
    "toString": [Function],
    "type": "NUM8",
    "value": "8",
  },
  {
    "col": 18,
    "line": 1,
    "lineBreaks": 0,
    "offset": 17,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 19,
    "text": "4",
    "toString": [Function],
    "type": "NUM4",
    "value": "4",
  },
  {
    "col": 22,
    "line": 1,
    "lineBreaks": 0,
    "offset": 21,
    "text": "@nthrt",
    "toString": [Function],
    "type": "NTHRT",
    "value": "@nthrt",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 28,
    "text": "8",
    "toString": [Function],
    "type": "NUM8",
    "value": "8",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 29,
    "text": "e",
    "toString": [Function],
    "type": "EXP10",
    "value": "e",
  },
  {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 30,
    "text": "5",
    "toString": [Function],
    "type": "NUM5",
    "value": "5",
  },
]
`;

exports[`tokenizer tokenize extracts tokens from the expression " 3+4 *$foo + sinh 1" 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "3",
    "toString": [Function],
    "type": "NUM3",
    "value": "3",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "+",
    "toString": [Function],
    "type": "POS",
    "value": "+",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "4",
    "toString": [Function],
    "type": "NUM4",
    "value": "4",
  },
  {
    "col": 6,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": "*",
    "toString": [Function],
    "type": "MUL",
    "value": "*",
  },
  {
    "col": 7,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": "$foo + sinh 1",
    "toString": [Function],
    "type": "syntaxError",
    "value": "$foo + sinh 1",
  },
]
`;

exports[`tokenizer tokenize extracts tokens from the expression " 3+4 *$foo + sinh PI" 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "3",
    "toString": [Function],
    "type": "NUM3",
    "value": "3",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "+",
    "toString": [Function],
    "type": "POS",
    "value": "+",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "4",
    "toString": [Function],
    "type": "NUM4",
    "value": "4",
  },
  {
    "col": 6,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": "*",
    "toString": [Function],
    "type": "MUL",
    "value": "*",
  },
  {
    "col": 7,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": "$",
    "toString": [Function],
    "type": "DOLLAR",
    "value": "$",
  },
  {
    "col": 8,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": "foo",
    "toString": [Function],
    "type": "FOO",
    "value": "foo",
  },
  {
    "col": 12,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 14,
    "line": 1,
    "lineBreaks": 0,
    "offset": 13,
    "text": "sinh",
    "toString": [Function],
    "type": "SINH",
    "value": "sinh",
  },
  {
    "col": 19,
    "line": 1,
    "lineBreaks": 0,
    "offset": 18,
    "text": "PI",
    "toString": [Function],
    "type": "PI",
    "value": "PI",
  },
]
`;

exports[`tokenizer tokenize extracts tokens from the expression "(.1 + .2) * 10^8 + 4 @nthrt 8e5" 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "(",
    "toString": [Function],
    "type": "LPAR",
    "value": "(",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": ".",
    "toString": [Function],
    "type": "DOT",
    "value": ".",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "1",
    "toString": [Function],
    "type": "NUM1",
    "value": "1",
  },
  {
    "col": 5,
    "line": 1,
    "lineBreaks": 0,
    "offset": 4,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 6,
    "text": ".",
    "toString": [Function],
    "type": "DOT",
    "value": ".",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": "2",
    "toString": [Function],
    "type": "NUM2",
    "value": "2",
  },
  {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": ")",
    "toString": [Function],
    "type": "RPAR",
    "value": ")",
  },
  {
    "col": 11,
    "line": 1,
    "lineBreaks": 0,
    "offset": 10,
    "text": "*",
    "toString": [Function],
    "type": "MUL",
    "value": "*",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 12,
    "text": "1",
    "toString": [Function],
    "type": "NUM1",
    "value": "1",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 13,
    "text": "0",
    "toString": [Function],
    "type": "NUM0",
    "value": "0",
  },
  {
    "col": 15,
    "line": 1,
    "lineBreaks": 0,
    "offset": 14,
    "text": "^",
    "toString": [Function],
    "type": "POW",
    "value": "^",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 15,
    "text": "8",
    "toString": [Function],
    "type": "NUM8",
    "value": "8",
  },
  {
    "col": 18,
    "line": 1,
    "lineBreaks": 0,
    "offset": 17,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 19,
    "text": "4",
    "toString": [Function],
    "type": "NUM4",
    "value": "4",
  },
  {
    "col": 22,
    "line": 1,
    "lineBreaks": 0,
    "offset": 21,
    "text": "@nthrt",
    "toString": [Function],
    "type": "NTHRT",
    "value": "@nthrt",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 28,
    "text": "8",
    "toString": [Function],
    "type": "NUM8",
    "value": "8",
  },
  {
    "col": 2,
    "line": 1,
    "lineBreaks": 0,
    "offset": 29,
    "text": "e",
    "toString": [Function],
    "type": "EXP10",
    "value": "e",
  },
  {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 30,
    "text": "5",
    "toString": [Function],
    "type": "NUM5",
    "value": "5",
  },
]
`;

exports[`tokenizer tokenize recognize all the registered terms 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 1,
    "text": "0",
    "toString": [Function],
    "type": "NUM0",
    "value": "0",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "1",
    "toString": [Function],
    "type": "NUM1",
    "value": "1",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 5,
    "text": "2",
    "toString": [Function],
    "type": "NUM2",
    "value": "2",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 7,
    "text": "3",
    "toString": [Function],
    "type": "NUM3",
    "value": "3",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": "4",
    "toString": [Function],
    "type": "NUM4",
    "value": "4",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 11,
    "text": "5",
    "toString": [Function],
    "type": "NUM5",
    "value": "5",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 13,
    "text": "6",
    "toString": [Function],
    "type": "NUM6",
    "value": "6",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 15,
    "text": "7",
    "toString": [Function],
    "type": "NUM7",
    "value": "7",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 17,
    "text": "8",
    "toString": [Function],
    "type": "NUM8",
    "value": "8",
  },
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 19,
    "text": "9",
    "toString": [Function],
    "type": "NUM9",
    "value": "9",
  },
  {
    "col": 22,
    "line": 1,
    "lineBreaks": 0,
    "offset": 21,
    "text": ".",
    "toString": [Function],
    "type": "DOT",
    "value": ".",
  },
  {
    "col": 24,
    "line": 1,
    "lineBreaks": 0,
    "offset": 23,
    "text": "e",
    "toString": [Function],
    "type": "EXP10",
    "value": "e",
  },
  {
    "col": 26,
    "line": 1,
    "lineBreaks": 0,
    "offset": 25,
    "text": "(",
    "toString": [Function],
    "type": "LPAR",
    "value": "(",
  },
  {
    "col": 28,
    "line": 1,
    "lineBreaks": 0,
    "offset": 27,
    "text": ")",
    "toString": [Function],
    "type": "RPAR",
    "value": ")",
  },
  {
    "col": 30,
    "line": 1,
    "lineBreaks": 0,
    "offset": 29,
    "text": ",",
    "toString": [Function],
    "type": "COMMA",
    "value": ",",
  },
  {
    "col": 32,
    "line": 1,
    "lineBreaks": 0,
    "offset": 31,
    "text": "~",
    "toString": [Function],
    "type": "ELLIPSIS",
    "value": "~",
  },
  {
    "col": 34,
    "line": 1,
    "lineBreaks": 0,
    "offset": 33,
    "text": "-",
    "toString": [Function],
    "type": "SUB",
    "value": "-",
  },
  {
    "col": 36,
    "line": 1,
    "lineBreaks": 0,
    "offset": 35,
    "text": "-",
    "toString": [Function],
    "type": "SUB",
    "value": "-",
  },
  {
    "col": 38,
    "line": 1,
    "lineBreaks": 0,
    "offset": 37,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 40,
    "line": 1,
    "lineBreaks": 0,
    "offset": 39,
    "text": "+",
    "toString": [Function],
    "type": "ADD",
    "value": "+",
  },
  {
    "col": 42,
    "line": 1,
    "lineBreaks": 0,
    "offset": 41,
    "text": "*",
    "toString": [Function],
    "type": "MUL",
    "value": "*",
  },
  {
    "col": 44,
    "line": 1,
    "lineBreaks": 0,
    "offset": 43,
    "text": "/",
    "toString": [Function],
    "type": "DIV",
    "value": "/",
  },
  {
    "col": 46,
    "line": 1,
    "lineBreaks": 0,
    "offset": 45,
    "text": "%",
    "toString": [Function],
    "type": "MOD",
    "value": "%",
  },
  {
    "col": 48,
    "line": 1,
    "lineBreaks": 0,
    "offset": 47,
    "text": "^",
    "toString": [Function],
    "type": "POW",
    "value": "^",
  },
  {
    "col": 50,
    "line": 1,
    "lineBreaks": 0,
    "offset": 49,
    "text": "!",
    "toString": [Function],
    "type": "FAC",
    "value": "!",
  },
  {
    "col": 52,
    "line": 1,
    "lineBreaks": 0,
    "offset": 51,
    "text": "=",
    "toString": [Function],
    "type": "ASSIGN",
    "value": "=",
  },
  {
    "col": 54,
    "line": 1,
    "lineBreaks": 0,
    "offset": 53,
    "text": "#",
    "toString": [Function],
    "type": "PERCENT",
    "value": "#",
  },
  {
    "col": 56,
    "line": 1,
    "lineBreaks": 0,
    "offset": 55,
    "text": "ans",
    "toString": [Function],
    "type": "VAR_ANS",
    "value": "ans",
  },
  {
    "col": 60,
    "line": 1,
    "lineBreaks": 0,
    "offset": 59,
    "text": "mem",
    "toString": [Function],
    "type": "VAR_MEM",
    "value": "mem",
  },
  {
    "col": 64,
    "line": 1,
    "lineBreaks": 0,
    "offset": 63,
    "text": "PI",
    "toString": [Function],
    "type": "PI",
    "value": "PI",
  },
  {
    "col": 67,
    "line": 1,
    "lineBreaks": 0,
    "offset": 66,
    "text": "E",
    "toString": [Function],
    "type": "E",
    "value": "E",
  },
  {
    "col": 69,
    "line": 1,
    "lineBreaks": 0,
    "offset": 68,
    "text": "TEN",
    "toString": [Function],
    "type": "TEN",
    "value": "TEN",
  },
  {
    "col": 73,
    "line": 1,
    "lineBreaks": 0,
    "offset": 72,
    "text": "NaN",
    "toString": [Function],
    "type": "NAN",
    "value": "NaN",
  },
  {
    "col": 77,
    "line": 1,
    "lineBreaks": 0,
    "offset": 76,
    "text": "Infinity",
    "toString": [Function],
    "type": "INFINITY",
    "value": "Infinity",
  },
  {
    "col": 86,
    "line": 1,
    "lineBreaks": 0,
    "offset": 85,
    "text": "Syntax",
    "toString": [Function],
    "type": "ERROR",
    "value": "Syntax",
  },
  {
    "col": 93,
    "line": 1,
    "lineBreaks": 0,
    "offset": 92,
    "text": "exp",
    "toString": [Function],
    "type": "EXP",
    "value": "exp",
  },
  {
    "col": 97,
    "line": 1,
    "lineBreaks": 0,
    "offset": 96,
    "text": "sqrt",
    "toString": [Function],
    "type": "SQRT",
    "value": "sqrt",
  },
  {
    "col": 102,
    "line": 1,
    "lineBreaks": 0,
    "offset": 101,
    "text": "cbrt",
    "toString": [Function],
    "type": "CBRT",
    "value": "cbrt",
  },
  {
    "col": 107,
    "line": 1,
    "lineBreaks": 0,
    "offset": 106,
    "text": "nthrt",
    "toString": [Function],
    "type": "NTHRT",
    "value": "nthrt",
  },
  {
    "col": 113,
    "line": 1,
    "lineBreaks": 0,
    "offset": 112,
    "text": "floor",
    "toString": [Function],
    "type": "FLOOR",
    "value": "floor",
  },
  {
    "col": 119,
    "line": 1,
    "lineBreaks": 0,
    "offset": 118,
    "text": "ceil",
    "toString": [Function],
    "type": "CEIL",
    "value": "ceil",
  },
  {
    "col": 124,
    "line": 1,
    "lineBreaks": 0,
    "offset": 123,
    "text": "round",
    "toString": [Function],
    "type": "ROUND",
    "value": "round",
  },
  {
    "col": 130,
    "line": 1,
    "lineBreaks": 0,
    "offset": 129,
    "text": "trunc",
    "toString": [Function],
    "type": "TRUNC",
    "value": "trunc",
  },
  {
    "col": 136,
    "line": 1,
    "lineBreaks": 0,
    "offset": 135,
    "text": "sin",
    "toString": [Function],
    "type": "SIN",
    "value": "sin",
  },
  {
    "col": 140,
    "line": 1,
    "lineBreaks": 0,
    "offset": 139,
    "text": "cos",
    "toString": [Function],
    "type": "COS",
    "value": "cos",
  },
  {
    "col": 144,
    "line": 1,
    "lineBreaks": 0,
    "offset": 143,
    "text": "tan",
    "toString": [Function],
    "type": "TAN",
    "value": "tan",
  },
  {
    "col": 148,
    "line": 1,
    "lineBreaks": 0,
    "offset": 147,
    "text": "asin",
    "toString": [Function],
    "type": "ASIN",
    "value": "asin",
  },
  {
    "col": 153,
    "line": 1,
    "lineBreaks": 0,
    "offset": 152,
    "text": "acos",
    "toString": [Function],
    "type": "ACOS",
    "value": "acos",
  },
  {
    "col": 158,
    "line": 1,
    "lineBreaks": 0,
    "offset": 157,
    "text": "atan",
    "toString": [Function],
    "type": "ATAN",
    "value": "atan",
  },
  {
    "col": 163,
    "line": 1,
    "lineBreaks": 0,
    "offset": 162,
    "text": "sinh",
    "toString": [Function],
    "type": "SINH",
    "value": "sinh",
  },
  {
    "col": 168,
    "line": 1,
    "lineBreaks": 0,
    "offset": 167,
    "text": "cosh",
    "toString": [Function],
    "type": "COSH",
    "value": "cosh",
  },
  {
    "col": 173,
    "line": 1,
    "lineBreaks": 0,
    "offset": 172,
    "text": "tanh",
    "toString": [Function],
    "type": "TANH",
    "value": "tanh",
  },
  {
    "col": 178,
    "line": 1,
    "lineBreaks": 0,
    "offset": 177,
    "text": "asinh",
    "toString": [Function],
    "type": "ASINH",
    "value": "asinh",
  },
  {
    "col": 184,
    "line": 1,
    "lineBreaks": 0,
    "offset": 183,
    "text": "acosh",
    "toString": [Function],
    "type": "ACOSH",
    "value": "acosh",
  },
  {
    "col": 190,
    "line": 1,
    "lineBreaks": 0,
    "offset": 189,
    "text": "atanh",
    "toString": [Function],
    "type": "ATANH",
    "value": "atanh",
  },
  {
    "col": 196,
    "line": 1,
    "lineBreaks": 0,
    "offset": 195,
    "text": "ln",
    "toString": [Function],
    "type": "LN",
    "value": "ln",
  },
  {
    "col": 199,
    "line": 1,
    "lineBreaks": 0,
    "offset": 198,
    "text": "log",
    "toString": [Function],
    "type": "LOG",
    "value": "log",
  },
  {
    "col": 203,
    "line": 1,
    "lineBreaks": 0,
    "offset": 202,
    "text": "lg",
    "toString": [Function],
    "type": "LG",
    "value": "lg",
  },
  {
    "col": 206,
    "line": 1,
    "lineBreaks": 0,
    "offset": 205,
    "text": "log10",
    "toString": [Function],
    "type": "LOG10",
    "value": "log10",
  },
  {
    "col": 212,
    "line": 1,
    "lineBreaks": 0,
    "offset": 211,
    "text": "abs",
    "toString": [Function],
    "type": "ABS",
    "value": "abs",
  },
  {
    "col": 216,
    "line": 1,
    "lineBreaks": 0,
    "offset": 215,
    "text": "random",
    "toString": [Function],
    "type": "RAND",
    "value": "random",
  },
]
`;
